File: Algorithms/DLX/Pascal/dlx.pas
program DLXSudoku;
{$mode objfpc}{$H+}
uses SysUtils;

type
  { Forward declarations }
  PDlxNode = ^TDlxNode;
  PDlxColumn = ^TDlxColumn;

  { DLX Node - basic node in the dancing links matrix }
  TDlxNode = record
    up, down, left, right: PDlxNode;
    column: PDlxColumn;
    row_id: Integer;
  end;

  { DLX Column - column header with size tracking }
  TDlxColumn = record
    node: TDlxNode;  { Inherits from DlxNode }
    size: Integer;
    name: String[16];
  end;

  { Row metadata to map DLX rows back to Sudoku (row, col, num) }
  TRowInfo = record
    row, col, num: Integer;
  end;

var
  { Global DLX structures }
  root: PDlxColumn;
  columns: array[0..323] of PDlxColumn;
  nodes: array of TDlxNode;
  node_count: Integer;
  row_info: array[0..728] of TRowInfo;
  row_starts: array[0..728] of PDlxNode;

  { Puzzle data }
  puzzle: array[0..8, 0..8] of Integer;
  solution_grid: array[0..8, 0..8] of Integer;

  { Iteration counter }
  dlx_iterations: Integer;

{ Calculate constraint column indices }
function GetPositionCol(r, c: Integer): Integer;
begin
  Result := r * 9 + c;
end;

function GetRowCol(r, n: Integer): Integer;
begin
  Result := 81 + r * 9 + (n - 1);
end;

function GetColCol(c, n: Integer): Integer;
begin
  Result := 162 + c * 9 + (n - 1);
end;

function GetBoxCol(r, c, n: Integer): Integer;
var
  box: Integer;
begin
  box := (r div 3) * 3 + (c div 3);
  Result := 243 + box * 9 + (n - 1);
end;

{ Initialize DLX matrix structure }
procedure InitDlxMatrix;
var
  i: Integer;
begin
  { Allocate root column }
  New(root);
  FillChar(root^, SizeOf(TDlxColumn), 0);
  root^.name := 'root';
  root^.node.left := @root^.node;
  root^.node.right := @root^.node;
  root^.node.up := @root^.node;
  root^.node.down := @root^.node;
  root^.node.column := root;
  root^.node.row_id := -1;
  root^.size := 0;

  { Allocate 324 column headers }
  for i := 0 to 323 do
  begin
    New(columns[i]);
    FillChar(columns[i]^, SizeOf(TDlxColumn), 0);
    columns[i]^.name := 'C' + IntToStr(i);
    columns[i]^.size := 0;

    { Initialize as circular list }
    columns[i]^.node.up := @columns[i]^.node;
    columns[i]^.node.down := @columns[i]^.node;
    columns[i]^.node.column := columns[i];
    columns[i]^.node.row_id := -1;

    { Link into header list }
    columns[i]^.node.left := root^.node.left;
    columns[i]^.node.right := @root^.node;
    root^.node.left^.right := @columns[i]^.node;
    root^.node.left := @columns[i]^.node;
  end;

  { Allocate node pool - 729 rows * 4 constraints each }
  SetLength(nodes, 729 * 4);
  node_count := 0;
end;

{ Add a node to the DLX matrix }
function AddNode(col: PDlxColumn; row_id: Integer): PDlxNode;
var
  node: PDlxNode;
begin
  if node_count >= Length(nodes) then
  begin
    WriteLn(StdErr, 'ERROR: Exceeded maximum node count');
    Halt(1);
  end;

  node := @nodes[node_count];
  Inc(node_count);

  node^.column := col;
  node^.row_id := row_id;

  { Insert at end of column's circular list }
  node^.down := @col^.node;
  node^.up := col^.node.up;
  col^.node.up^.down := node;
  col^.node.up := node;
  Inc(col^.size);

  Result := node;
end;

{ Build a DLX row for Sudoku cell (r,c) with value n }
procedure BuildDlxRow(r, c, n, row_id: Integer);
var
  n1, n2, n3, n4: PDlxNode;
begin
  { Store row metadata }
  row_info[row_id].row := r;
  row_info[row_id].col := c;
  row_info[row_id].num := n;

  { Create nodes for the 4 constraints }
  n1 := AddNode(columns[GetPositionCol(r, c)], row_id);
  n2 := AddNode(columns[GetRowCol(r, n)], row_id);
  n3 := AddNode(columns[GetColCol(c, n)], row_id);
  n4 := AddNode(columns[GetBoxCol(r, c, n)], row_id);

  { Link nodes horizontally in circular list }
  n1^.right := n2;
  n2^.right := n3;
  n3^.right := n4;
  n4^.right := n1;

  n1^.left := n4;
  n2^.left := n1;
  n3^.left := n2;
  n4^.left := n3;

  { Store first node for this row }
  row_starts[row_id] := n1;
end;

{ Build the complete DLX matrix from the puzzle }
procedure BuildDlxMatrixFromPuzzle;
var
  r, c, n, row_id: Integer;
begin
  row_id := 0;

  for r := 0 to 8 do
  begin
    for c := 0 to 8 do
    begin
      if puzzle[r, c] <> 0 then
      begin
        { Cell has a clue - create only one row for that value }
        BuildDlxRow(r, c, puzzle[r, c], row_id);
        Inc(row_id);
      end
      else
      begin
        { Cell is empty - create rows for all possible values }
        for n := 1 to 9 do
        begin
          BuildDlxRow(r, c, n, row_id);
          Inc(row_id);
        end;
      end;
    end;
  end;
end;

{ Cover a column in the DLX matrix }
procedure DlxCoverColumn(c: PDlxColumn);
var
  col_node, row_node, right_node: PDlxNode;
begin
  col_node := @c^.node;

  { Remove column header from the header list }
  col_node^.right^.left := col_node^.left;
  col_node^.left^.right := col_node^.right;

  { For each row in this column }
  row_node := col_node^.down;
  while row_node <> col_node do
  begin
    { For each node in this row (excluding the column itself) }
    right_node := row_node^.right;
    while right_node <> row_node do
    begin
      { Remove this node from its column }
      right_node^.down^.up := right_node^.up;
      right_node^.up^.down := right_node^.down;
      Dec(right_node^.column^.size);
      right_node := right_node^.right;
    end;
    row_node := row_node^.down;
  end;
end;

{ Uncover a column (exact reverse of cover) }
procedure DlxUncoverColumn(c: PDlxColumn);
var
  col_node, row_node, left_node: PDlxNode;
begin
  col_node := @c^.node;

  { For each row in this column (in reverse order) }
  row_node := col_node^.up;
  while row_node <> col_node do
  begin
    { For each node in this row (in reverse order) }
    left_node := row_node^.left;
    while left_node <> row_node do
    begin
      { Restore this node to its column }
      Inc(left_node^.column^.size);
      left_node^.down^.up := left_node;
      left_node^.up^.down := left_node;
      left_node := left_node^.left;
    end;
    row_node := row_node^.up;
  end;

  { Restore column header to the header list }
  col_node^.right^.left := col_node;
  col_node^.left^.right := col_node;
end;

{ Choose column with minimum size (Knuth's S heuristic) }
function ChooseColumn(root: PDlxColumn): PDlxColumn;
var
  root_node, col_node: PDlxNode;
  best: PDlxColumn;
  min_size: Integer;
  col: PDlxColumn;
begin
  root_node := @root^.node;
  best := nil;
  min_size := High(Integer);

  col_node := root_node^.right;
  while col_node <> root_node do
  begin
    col := PDlxColumn(col_node);
    if col^.size < min_size then
    begin
      min_size := col^.size;
      best := col;
    end;
    col_node := col_node^.right;
  end;

  Result := best;
end;

{ DLX Search - Algorithm X with Dancing Links }
function DlxSearch(root: PDlxColumn; k: Integer; var solution: array of Integer): Boolean;
var
  root_node, row_node, right_node, left_node: PDlxNode;
  col: PDlxColumn;
begin
  { Count every search call }
  Inc(dlx_iterations);

  root_node := @root^.node;

  { If matrix is empty, we found a solution }
  if root_node^.right = root_node then
    Exit(True);

  { Choose column with minimum size }
  col := ChooseColumn(root);

  { If column has no rows, no solution possible }
  if col^.size = 0 then
    Exit(False);

  { Cover this column }
  DlxCoverColumn(col);

  { Try each row in this column }
  row_node := col^.node.down;
  while row_node <> @col^.node do
  begin
    { Add row to partial solution }
    solution[k] := row_node^.row_id;

    { Cover all other columns in this row }
    right_node := row_node^.right;
    while right_node <> row_node do
    begin
      DlxCoverColumn(right_node^.column);
      right_node := right_node^.right;
    end;

    { Recurse }
    if DlxSearch(root, k + 1, solution) then
      Exit(True);

    { Backtrack: uncover all columns in this row }
    left_node := row_node^.left;
    while left_node <> row_node do
    begin
      DlxUncoverColumn(left_node^.column);
      left_node := left_node^.left;
    end;

    row_node := row_node^.down;
  end;

  { Uncover column }
  DlxUncoverColumn(col);

  Result := False;
end;

{ Cover given clues (pre-selected rows) }
procedure CoverClues;
var
  r, c, n, row_id: Integer;
  node, curr: PDlxNode;
begin
  for r := 0 to 8 do
  begin
    for c := 0 to 8 do
    begin
      if puzzle[r, c] <> 0 then
      begin
        n := puzzle[r, c];

        { Find the row for this clue }
        for row_id := 0 to 728 do
        begin
          if (row_starts[row_id] <> nil) and
             (row_info[row_id].row = r) and
             (row_info[row_id].col = c) and
             (row_info[row_id].num = n) then
          begin
            { Cover all columns in this row }
            node := row_starts[row_id];
            curr := node;
            repeat
              DlxCoverColumn(curr^.column);
              curr := curr^.right;
            until curr = node;
            Break;
          end;
        end;
      end;
    end;
  end;
end;

{ Extract solution from DLX and populate solution_grid }
procedure ExtractSolution(var solution: array of Integer; solution_len: Integer);
var
  i, row_id: Integer;
begin
  { Initialize solution grid - start with the original puzzle }
  solution_grid := puzzle;

  { Each solution entry is a row_id }
  for i := 0 to solution_len - 1 do
  begin
    row_id := solution[i];
    if (row_id >= 0) and (row_id < 729) then
    begin
      solution_grid[row_info[row_id].row, row_info[row_id].col] := row_info[row_id].num;
    end;
  end;
end;

{ Print puzzle }
procedure PrintPuzzle;
var
  r, c: Integer;
begin
  WriteLn;
  WriteLn('Puzzle:');
  for r := 0 to 8 do
  begin
    for c := 0 to 8 do
      Write(puzzle[r, c], ' ');
    WriteLn;
  end;
end;

procedure PrintSolution;
var
  r, c: Integer;
begin
  WriteLn;
  WriteLn('Puzzle:');
  for r := 0 to 8 do
  begin
    for c := 0 to 8 do
      Write(solution_grid[r, c], ' ');
    WriteLn;
  end;
end;

{ Read matrix file }
function ReadMatrixFile(const filename: string): Boolean;
var
  f: TextFile;
  line: string;
  r, c, idx: Integer;
  display_path: string;
begin
  Result := False;

  { Print filename (normalize /app/Matrices to ../Matrices) }
  if Pos('/app/Matrices/', filename) = 1 then
  begin
    display_path := '../' + Copy(filename, 6, Length(filename));
    WriteLn(display_path);
  end
  else
    WriteLn(filename);

  AssignFile(f, filename);
  try
    Reset(f);
  except
    WriteLn(StdErr, 'Error opening file: ', filename);
    Exit;
  end;

  r := 0;
  while (not Eof(f)) and (r < 9) do
  begin
    ReadLn(f, line);
    line := Trim(line);

    { Skip comments and empty lines }
    if (line = '') then
      Continue;
    if (Length(line) > 0) and (line[1] = '#') then
      Continue;

    { Parse 9 space-separated integers }
    c := 0;
    idx := 1;
    while (idx <= Length(line)) and (c < 9) do
    begin
      { Skip spaces }
      while (idx <= Length(line)) and (line[idx] = ' ') do
        Inc(idx);

      if idx <= Length(line) then
      begin
        puzzle[r, c] := Ord(line[idx]) - Ord('0');
        Write(puzzle[r, c], ' ');
        Inc(c);
        Inc(idx);
      end;
    end;
    WriteLn;
    Inc(r);
  end;

  CloseFile(f);
  Result := True;
end;

{ Free DLX matrix memory }
procedure FreeDlxMatrix;
var
  i: Integer;
begin
  if root <> nil then
    Dispose(root);
  for i := 0 to 323 do
    if columns[i] <> nil then
      Dispose(columns[i]);
  SetLength(nodes, 0);
end;

{ Main program }
var
  i: Integer;
  filename: string;
  solution: array[0..80] of Integer;
  result: Boolean;
  start_time, end_time: TDateTime;
begin
  if ParamCount < 1 then
  begin
    WriteLn(StdErr, 'Usage: dlx <matrix_file>');
    Halt(1);
  end;

  start_time := Now;

  for i := 1 to ParamCount do
  begin
    filename := ParamStr(i);
    if Pos('.matrix', filename) > 0 then
    begin
      if ReadMatrixFile(filename) then
      begin
        PrintPuzzle;

        { Initialize DLX matrix }
        InitDlxMatrix;

        { Build matrix from puzzle }
        BuildDlxMatrixFromPuzzle;

        { Cover pre-filled clues }
        CoverClues;

        { Solve using DLX }
        dlx_iterations := 0;
        FillChar(solution, SizeOf(solution), 0);
        result := DlxSearch(root, 0, solution);

        if result then
        begin
          ExtractSolution(solution, 81);
          PrintSolution;
          WriteLn;
          WriteLn('Solved in Iterations=', dlx_iterations);
          WriteLn;
        end
        else
          WriteLn('No solution found');

        { Cleanup }
        FreeDlxMatrix;
      end;
    end;
  end;

  end_time := Now;
  WriteLn(Format('Seconds to process %.3f', [(end_time - start_time) * 86400]));
end.

File: Algorithms/DLX/Swift/dlx.swift
import Foundation

// DLX Node class - uses reference semantics for circular linked structure
class DlxNode {
    var left: DlxNode?
    var right: DlxNode?
    var up: DlxNode?
    var down: DlxNode?
    var column: DlxNode?  // Points to column header
    var size: Int = 0     // Only used by column headers
    var rowId: Int = -1   // Row ID for solution tracking
    var name: String = ""

    init() {}
}

// Global iteration counter
var dlxIterations = 0

// Cover a column in the DLX matrix
func coverColumn(_ col: DlxNode) {
    // Remove column header from the header list
    col.right?.left = col.left
    col.left?.right = col.right

    // For each row in this column
    var rowNode = col.down
    while rowNode !== col {
        // For each node in this row (excluding the column itself)
        var rightNode = rowNode?.right
        while rightNode !== rowNode {
            // Remove this node from its column
            rightNode?.down?.up = rightNode?.up
            rightNode?.up?.down = rightNode?.down
            if let colNode = rightNode?.column {
                colNode.size -= 1
            }
            rightNode = rightNode?.right
        }
        rowNode = rowNode?.down
    }
}

// Uncover a column (exact reverse of cover)
func uncoverColumn(_ col: DlxNode) {
    // For each row in this column (in reverse order)
    var rowNode = col.up
    while rowNode !== col {
        // For each node in this row (in reverse order)
        var leftNode = rowNode?.left
        while leftNode !== rowNode {
            // Restore this node to its column
            if let colNode = leftNode?.column {
                colNode.size += 1
            }
            leftNode?.down?.up = leftNode
            leftNode?.up?.down = leftNode
            leftNode = leftNode?.left
        }
        rowNode = rowNode?.up
    }

    // Restore column header to the header list
    col.right?.left = col
    col.left?.right = col
}

// Choose column with minimum size (Knuth's S heuristic)
func chooseColumn(_ root: DlxNode) -> DlxNode? {
    var best: DlxNode? = nil
    var minSize = Int.max

    var colNode = root.right
    while colNode !== root {
        if let col = colNode {
            if col.size < minSize {
                minSize = col.size
                best = col
            }
        }
        colNode = colNode?.right
    }

    return best
}

// DLX Search - Algorithm X with Dancing Links
func dlxSearch(_ root: DlxNode, _ k: Int, _ solution: inout [Int]) -> Bool {
    dlxIterations += 1  // Count every search call

    // If matrix is empty, we found a solution
    if root.right === root {
        return true
    }

    // Choose column with minimum size
    guard let col = chooseColumn(root) else {
        return false
    }

    // If column has no rows, no solution possible
    if col.size == 0 {
        return false
    }

    // Cover this column
    coverColumn(col)

    // Try each row in this column
    var rowNode = col.down
    while rowNode !== col {
        if let row = rowNode {
            // Add row to partial solution
            solution[k] = row.rowId

            // Cover all other columns in this row
            var rightNode = row.right
            while rightNode !== row {
                if let rn = rightNode, let colTocover = rn.column {
                    coverColumn(colTocover)
                }
                rightNode = rightNode?.right
            }

            // Recurse
            if dlxSearch(root, k + 1, &solution) {
                return true  // Solution found
            }

            // Backtrack: uncover all columns in this row
            var leftNode = row.left
            while leftNode !== row {
                if let ln = leftNode, let colToUncover = ln.column {
                    uncoverColumn(colToUncover)
                }
                leftNode = leftNode?.left
            }
        }

        rowNode = rowNode?.down
    }

    // Uncover column
    uncoverColumn(col)

    return false  // No solution found
}

// Sudoku-specific structures
struct RowInfo {
    var row: Int
    var col: Int
    var num: Int
}

var puzzle = Array(repeating: Array(repeating: 0, count: 9), count: 9)
var rowInfo = Array(repeating: RowInfo(row: 0, col: 0, num: 0), count: 729)
var rowStarts = Array<DlxNode?>(repeating: nil, count: 729)

// Calculate constraint column indices
func getPositionCol(_ r: Int, _ c: Int) -> Int {
    return r * 9 + c
}

func getRowCol(_ r: Int, _ n: Int) -> Int {
    return 81 + r * 9 + (n - 1)
}

func getColCol(_ c: Int, _ n: Int) -> Int {
    return 162 + c * 9 + (n - 1)
}

func getBoxCol(_ r: Int, _ c: Int, _ n: Int) -> Int {
    let box = (r / 3) * 3 + (c / 3)
    return 243 + box * 9 + (n - 1)
}

// Add a node to the DLX matrix
func addNode(_ col: DlxNode, _ rowId: Int, _ allNodes: inout [DlxNode]) -> DlxNode {
    let node = DlxNode()
    allNodes.append(node)

    node.column = col
    node.rowId = rowId

    // Insert at end of column's circular list
    node.down = col
    node.up = col.up
    col.up?.down = node
    col.up = node
    col.size += 1

    return node
}

// Build a DLX row for Sudoku cell (r,c) with value n
func buildDlxRow(_ r: Int, _ c: Int, _ n: Int, _ rowId: Int,
                 _ columns: [DlxNode], _ allNodes: inout [DlxNode]) {
    // Store row metadata
    rowInfo[rowId] = RowInfo(row: r, col: c, num: n)

    // Create nodes for the 4 constraints
    let n1 = addNode(columns[getPositionCol(r, c)], rowId, &allNodes)
    let n2 = addNode(columns[getRowCol(r, n)], rowId, &allNodes)
    let n3 = addNode(columns[getColCol(c, n)], rowId, &allNodes)
    let n4 = addNode(columns[getBoxCol(r, c, n)], rowId, &allNodes)

    // Link nodes horizontally in circular list
    n1.right = n2
    n2.right = n3
    n3.right = n4
    n4.right = n1

    n1.left = n4
    n2.left = n1
    n3.left = n2
    n4.left = n3

    // Store first node for this row
    rowStarts[rowId] = n1
}

// Initialize DLX matrix structure
func initDlxMatrix() -> (DlxNode, [DlxNode]) {
    // Create root column
    let root = DlxNode()
    root.name = "root"
    root.left = root
    root.right = root
    root.up = root
    root.down = root
    root.column = root
    root.rowId = -1
    root.size = 0

    // Create 324 column headers
    var columns: [DlxNode] = []
    for i in 0..<324 {
        let col = DlxNode()
        col.name = "C\(i)"
        col.size = 0
        col.up = col
        col.down = col
        col.column = col
        col.rowId = -1

        // Link into header list
        col.left = root.left
        col.right = root
        root.left?.right = col
        root.left = col

        columns.append(col)
    }

    return (root, columns)
}

// Build the complete DLX matrix from the puzzle
func buildDlxMatrixFromPuzzle(_ columns: [DlxNode], _ allNodes: inout [DlxNode]) {
    var rowId = 0

    for r in 0..<9 {
        for c in 0..<9 {
            if puzzle[r][c] != 0 {
                // Cell has a clue - create only one row for that value
                buildDlxRow(r, c, puzzle[r][c], rowId, columns, &allNodes)
                rowId += 1
            } else {
                // Cell is empty - create rows for all possible values
                for n in 1...9 {
                    buildDlxRow(r, c, n, rowId, columns, &allNodes)
                    rowId += 1
                }
            }
        }
    }
}

// Cover given clues (pre-selected rows)
func coverClues(_ root: DlxNode) {
    for r in 0..<9 {
        for c in 0..<9 {
            if puzzle[r][c] != 0 {
                let n = puzzle[r][c]

                // Find the row for this clue
                for rowId in 0..<729 {
                    if let node = rowStarts[rowId] {
                        if rowInfo[rowId].row == r &&
                           rowInfo[rowId].col == c &&
                           rowInfo[rowId].num == n {
                            // Cover all columns in this row
                            var curr = node
                            repeat {
                                if let colToCover = curr.column {
                                    coverColumn(colToCover)
                                }
                                curr = curr.right!
                            } while curr !== node
                            break
                        }
                    }
                }
            }
        }
    }
}

// Extract solution from DLX
func extractSolution(_ solution: [Int]) -> [[Int]] {
    var solutionGrid = puzzle

    for i in 0..<81 {
        let rowId = solution[i]
        if rowId >= 0 && rowId < 729 {
            solutionGrid[rowInfo[rowId].row][rowInfo[rowId].col] = rowInfo[rowId].num
        }
    }

    return solutionGrid
}

// Print puzzle
func printPuzzle(_ grid: [[Int]]) {
    print("\nPuzzle:")
    for r in 0..<9 {
        for c in 0..<9 {
            print("\(grid[r][c]) ", terminator: "")
        }
        print("")
    }
}

// Read matrix file
func readMatrixFile(_ filename: String) -> Bool {
    guard let content = try? String(contentsOfFile: filename, encoding: .utf8) else {
        print("Error reading file \(filename)")
        return false
    }

    // Normalize path for output
    let displayPath: String
    if filename.hasPrefix("/app/Matrices/") {
        displayPath = "../" + String(filename.dropFirst(5))
    } else {
        displayPath = filename
    }
    print(displayPath)

    let lines = content.components(separatedBy: .newlines)
    var row = 0

    for line in lines {
        if row >= 9 { break }

        let trimmed = line.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty || trimmed.hasPrefix("#") { continue }

        let parts = trimmed.components(separatedBy: .whitespaces).filter { !$0.isEmpty }

        if parts.count >= 9 {
            for col in 0..<9 {
                if let val = Int(parts[col]) {
                    puzzle[row][col] = val
                    print("\(val) ", terminator: "")
                }
            }
            print("")
            row += 1
        }
    }

    return row == 9
}

// Main execution
let args = CommandLine.arguments
if args.count < 2 {
    print("Usage: dlx_solver <matrix_file>")
    exit(1)
}

let startTime = Date()

// Process each matrix file
for i in 1..<args.count {
    let filename = args[i]

    if filename.hasSuffix(".matrix") {
        // Reset state
        dlxIterations = 0
        puzzle = Array(repeating: Array(repeating: 0, count: 9), count: 9)
        rowInfo = Array(repeating: RowInfo(row: 0, col: 0, num: 0), count: 729)
        rowStarts = Array<DlxNode?>(repeating: nil, count: 729)

        if readMatrixFile(filename) {
            printPuzzle(puzzle)

            // Initialize DLX matrix
            let (root, columns) = initDlxMatrix()
            var allNodes: [DlxNode] = []

            // Build matrix from puzzle
            buildDlxMatrixFromPuzzle(columns, &allNodes)

            // Cover pre-filled clues
            coverClues(root)

            // Solve using DLX
            var solution = Array(repeating: -1, count: 81)
            let result = dlxSearch(root, 0, &solution)

            if result {
                let solutionGrid = extractSolution(solution)
                printPuzzle(solutionGrid)
                print("\nSolved in Iterations=\(dlxIterations)\n")
            } else {
                print("\nNo solution found")
            }
        }
    }
}

let endTime = Date()
let elapsed = endTime.timeIntervalSince(startTime)
print(String(format: "Seconds to process %.3f", elapsed))
